Redirects and security considerations (#143 and #112) (#146)
securitytxt/security-txt: draft-foudil-securitytxt.md

# Location of the security.txt file

## Web-based services {#weblocation}

Web-based services SHOULD place the security.txt file under the /.well-known/ path; e.g. https://example.com/.well-known/security.txt
as per {{!RFC5785}}. A security.txt file located under the top-level path SHOULD either redirect (as per section 6.4 of {{!RFC7231}})

> to the security.txt file under the /.well-known/ path or be used as a fall back if the ".well-known" path cannot be used.

Author: fixing

If retrieval of a "security.txt" file results in a redirect (as per
section 6.4 of {{!RFC7231}}), the implementors MUST NOT follow
redirects that lead to another domain or subdomain

> but MAY follow redirects within the same domain name

Author: fixing

"ext-field" refers to extension fields, which are discussed in {{extensibility}}

# Security Considerations

## Compromised Files and Redirects

An attacker that has compromised a website is able to compromise
the "security.txt" file as well or setup a redirect to their own site.
This can result in security reports not being received by the organization
or sent to the attacker.

To protect against this, organizations SHOULD digitally sign their "security.txt"
files (as per {{signature}}), use the canonical directive to sign the location
of the file (as per {{canonical}}), and regularly monitor the file and
the referenced resources to detect tampering.

> Security researchers SHOULD check verify the "security.txt" including the

Author: fixing

> the digital signature and checking historical records before using the information

Reviewer: By "historical records", do you mean stuff like the Wayback machine? Or records that the security researcher is keeping themselves?

Author: either

the information in a "security.txt" file looks suspicious or compromised,
it SHOULD NOT be used.

To avoid redirect attacks, redirects for these files MUST NOT be followed
if they lead to a different domain (as per {{weblocation}}).

## Incorrect or Stale Information

If information and resources referenced in a "security.txt" file are incorrect
or not kept up to date, this can result in security reports not being received
by the organization or sent to incorrect contacts, thus exposing possible
security issues to third parties.

Organizations SHOULD ensure that any this file and any resources such as web pages,

> email addresses and telephone numbers references by a "security.txt"

Author: fixing

are kept current, are accessible, controlled by the organization,
and are kept secure.

## Intentionally Malformed Files, Resources and Reports

It is possible for attackers to generate files that are extraordinarily
large or otherwise malformed in an attempt to discover or exploit weaknesses
in parsing code. Implementors SHOULD make sure that any such code
be robust against large and malformed files. ABNF grammar (as defined in
{{abnf}}) SHOULD be used a way to verify these files.

Same concerns apply to any other resources referenced within security.txt

> files, as well as any security reports received as the result of publishing

Reviewer 2: ```suggestion
files, as well as any security reports received as a result of publishing
```

this file. Such resources and reports may be hostile, malformed or malicious.

## No Implied Permission for Testing

The presence of a security.txt file can be interpreted as providing permission
to do security testing against that asset. This can lead to increased testing
against an organization by researchers. On the other hand, a decision not

> to publish a security.txt file can be interpreted by an organization as

Reviewer: I find "can be interpreted by an organization" a little confusing because isn't it the researchers which would be doing the interpreting? Or is it the organisation that's choosing to not publish a security.txt file doing the interpreting (in which case, consider changing "an" to "the")?

Author: clarifying language

> to publish a security.txt file on a particular website, can be interpreted by the

Reviewer: "on a particular website" excludes ".security.txt" files (placed at the root of the source code of software), which also has a similar risk about attracting spam.

Author: fixed language

> organization operating that website as a way to signal to researchers

Reviewer: I think this sounds better:

```suggestion
organization operating that website to be a way to signal to researchers
```

Author: fixed

This can lead to pushback against researchers reporting security issues

> to organization.

Reviewer: Should this be "organizations" (plural)?

Author: clarifying

Therefore implementors MUST NOT assume that presence or absence

> of a security.txt files grants or denies permission for security testing.

Reviewer: "a security.txt files" -- consider making "files" singular to match with the article:

```suggestion
of a security.txt file grants or denies permission for security testing.
```

> Therefore implementors MUST NOT assume that presence or absence of

Reviewer 2: ```suggestion
Therefore, implementors MUST NOT assume that the presence or absence of
```

Any such permission MAY be defined in a security or disclosure policy
(as per {{policy}}) or a new directive (as per {{extensibility}}).

## Multi-user Environments

In multi-user / multi-tenant environments, it may possible for a user to take
over the location of the "security.txt" file. Organizations SHOULD reserve
the "security.txt" namespace to ensure no third-party can create a page with
the "security.txt" AND "/.well-known/security.txt" names.

## Protecting Data in Transit

> To protect security.txt file from being tampered in transit, implementors MUST use

Author: fixing

> HTTPS for the file itself and any web URLs referenced in it (except as noted in

Reviewer 2: ```suggestion
HTTPS when serving the file itself and any web URLs referenced in it (except as noted in
```

this specification). Implementors MUST also perform the correct TLS
verification (as per {{!RFC6125}}).

> As an additional layer of protection. it is also RECOMMENDED that

Reviewer 2: ```suggestion
As an additional layer of protection, it is also RECOMMENDED that
```

> organizations digitally sign this file with OpenPGP (as per {{signature}}).

Reviewer 2: ```suggestion
organizations digitally sign their "security.txt" file with OpenPGP (as per {{signature}}).
```

> Also, to protect security reports from being tampered or seen while in transit,

Reviewer 2: ```suggestion
Also, to protect security reports from being tampered with or seen while in transit,
```

by the organization or sent to incorrect contacts, thus exposing possible
security issues to third parties.

Organizations SHOULD ensure that information in this file and any referenced
resources such as web pages, email addresses and telephone numbers
are kept current, are accessible, controlled by the organization,
and are kept secure.

## Intentionally Malformed Files, Resources and Reports

It is possible for attackers to generate files that are extraordinarily
large or otherwise malformed in an attempt to discover or exploit weaknesses
in parsing code. Implementors SHOULD make sure that any such code
be robust against large and malformed files. ABNF grammar (as defined in
{{abnf}}) SHOULD be used a way to verify these files.

Same concerns apply to any other resources referenced within security.txt
files, as well as any security reports received as the result of publishing
this file. Such resources and reports may be hostile, malformed or malicious.

## No Implied Permission for Testing

The presence of a security.txt file can be interpreted by researchers
as providing permission to do security testing against that asset.

> this can lead to increased testing

Reviewer: ```suggestion
This can lead to increased testing
```

also I think the line has been broken a little too early?

## Since draft-foudil-securitytxt-02
- Use "mailto" and "tel" (#62)
- Fix typo in the "Example" section (#64)

> - Clarified that the root directory is a fallback option (#72)

Reviewer: I have no experience with writing RFCs and the like, so wanted to just make sure that changing the history to fix a typo is okay. It seems like a reasonable thing to do but I just want to make sure.

Author: That's usually fine, this section is not normative

## Since draft-foudil-securitytxt-05
- Changing HTTPS to MUST (#55)
- Adding language recommending encryption for email reports (#134)
- Added language handling redirects (#143)

> - Expanded security considerations section (#30, #73, #103, #112)

Reviewer: Not sure if it's notable enough but you also fixed some typos with "fall back" -> "fallback". Spelling corrections were mentioned in the history update for the "since -02" entry.
